### PR TITLE
Feature/base metric

### PR DESCRIPTION
**Pull request by uduse**
#312

### CODE REVIEW

File: matchzoo/engine/base_model.py [hunk 1]
```diff
@@ -186,8 +189,31 @@ def evaluate(
...
+                                               verbose=verbose)
+        if not isinstance(backend_evals, list):
+            backend_evals = [backend_evals]
+        metrics_lookup = {name: val for name, val in
+                          zip(self._backend.metrics_names, backend_evals)}
+        list_wise_df = None
+        for metric in self._params['task'].metrics:
+            if isinstance(metric, engine.BaseMetric):
+                if list_wise_df is None:
+                    y_pred = self.predict(x, batch_size)
```
Comment (faneshion) [inline matchzoo/engine/base_model.py]: Here, we can adjust the code to improve efficiency. The line `y_pred = self.predict(x, batch_size)` can be moved out of the loop `for'.

File: matchzoo/engine/base_task.py [hunk 2]
```diff
@@ -3,10 +3,49 @@
 import typing
 import abc
 
+from matchzoo import engine
+
 
 class BaseTask(abc.ABC):
     """Base Task, shouldn't be used directly."""
 
+    @classmethod
```
Comment (bwanglzu) [inline matchzoo/engine/base_task.py]: why @classmethod

File: matchzoo/metrics.py [hunk 3]
```diff
@@ -1,190 +1,254 @@
...
+        for idx, (label, score) in enumerate(coupled_pair):
+            if idx >= self._k:
+                break
+            if label > self._threshold:
+                precision += 1.
+        return precision / self._k
+
+
+class AveragePrecision(engine.BaseMetric):
+    ALIAS = 'average_precision'
```
Comment (bwanglzu) [inline matchzoo/metrics.py]: ['average_precision', 'ap']

File: matchzoo/metrics.py [hunk 4]
```diff
@@ -1,190 +1,254 @@
...
+    labels = np.squeeze(labels).tolist()
+    scores = np.squeeze(scores).tolist()
+    couple = list(zip(labels, scores))
+    random.shuffle(couple)
+    sorted_couple = sorted(couple, key=lambda x: x[1], reverse=True)
+    return sorted_couple
+
+
+class MeanReciprocalRank(engine.BaseMetric):
+    ALIAS = ['mean_reciprocal_rank', 'mrr', 'MRR']
```
Comment (bwanglzu) [inline matchzoo/metrics.py]: shall we create a function in base class to convert alias to lowercase so that we do not necessarily need to write   'mrr'  and 'MRR' (also for other sub-classes).
It's a bit tricky since some times people use `normalized_discounted_cumulative_gain`, sometimes `ndcg`, `NDCG` or even `nDCG`

File: matchzoo/metrics.py [hunk 5]
```diff
@@ -1,190 +1,254 @@
...
+        if pos == 0:
+            return 0.
+        else:
+            return result / pos
+
+
+class DiscountedCumulativeGain(engine.BaseMetric):
+    ALIAS = ['discounted_cumulative_gain', 'dcg', 'DCG']
+
+    def __init__(self, k=1, threshold=0):
```
Comment (bwanglzu) [inline matchzoo/metrics.py]: change the default `k` to 10? also, `k = 10` and `threshold = 0`
`pip install flake8 --upgrade`

File: matchzoo/engine/base_model.py [hunk 6]
```diff
@@ -105,9 +104,13 @@ def build(self):
 
     def compile(self):
         """Compile model for training."""
+        keras_metrics = []
+        for metric in self._params['task'].metrics:
+            if not isinstance(engine.parse_metric(metric), engine.BaseMetric):
+                keras_metrics.append(metric)
```
Comment (bwanglzu) [inline matchzoo/engine/base_model.py]: I guess a better way to do is:
```python
activated_keras_metrics = []
for metric in self._params['task'].metrics:
    if keras.metrics.get(metric):
        activated_keras_metrics.append(metric)
```
take a look at [here](https://github.com/keras-team/keras/blob/df8186833fe47a6fd4b7dbc52fe12f9df73609fb/keras/metrics.py#L76), for the sake of parse.

**Comment by bwanglzu** [comment]
`git merge 2.0` and you'll get all pep8 stuff fixed

File: matchzoo/metrics.py (hunk 4 above)
Comment (faneshion) [reply matchzoo/metrics.py]: That's right~

File: matchzoo/engine/base_model.py (hunk 1 above)
Comment (uduse) [reply matchzoo/engine/base_model.py]: It only happens once already.

File: matchzoo/engine/base_task.py (hunk 2 above)
Comment (uduse) [reply matchzoo/engine/base_task.py]: Why not?

File: matchzoo/engine/base_model.py (hunk 6 above)
Comment (uduse) [reply matchzoo/engine/base_model.py]: A metric is one of keras' metric, matchzoo's metric, or a erroneous metric (e.g. a metric with a typo). I propagate erroneous metrics to keras so keras will throw exceptions for me.

File: matchzoo/metrics.py (hunk 5 above)
Comment (uduse) [reply matchzoo/metrics.py]: @faneshion had 1 but I don't know why.

File: matchzoo/metrics.py (hunk 3 above)
Comment (uduse) [reply matchzoo/metrics.py]: Fixed

File: matchzoo/metrics.py (hunk 4 above)
Comment (uduse) [reply matchzoo/metrics.py]: Fixed
